### ISSUE TITLE
Add no-unknown-custom-properties rule

### ISSUE THREAD

**Issue by dylang** [stylelint/stylelint#2220]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

New rule: Check that all css variables have been declared.

> Which rule, if any, is this issue related to?

I think this may need to be a new rule.

> What CSS is needed to reproduce this issue?

```css
:root {
    --colorBlue: #3e92e6;
}
```

```css
:root {
    --colorPrimary: var(--colorBlue); // no problem
    --colorUndefined: var(--colorNeverBeenSeenBefore); // oh no!
}
```

> What stylelint configuration is needed to reproduce this issue?

n/a

> Which version of stylelint are you using?

`7.7.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

cli

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

CSS: https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_variables

> What did you expect to happen?

```shell
test.css
1:21    Unknown variable colorNeverBeenSeenBefore (no-unknown-variables)
```

> What actually happened (e.g. what warnings or errors you are getting)?

"No warnings to be flagged."

**Comment by evilebottnawi** [comment]
@dylang Good rule :+1:  @stylelint/core what do you think about the support new rules in core (`custom-selectors`, `variables` and other from new spec)

**Comment by jeddy3** [comment]
@dylang Thanks for the suggestion and for using the template. I agree that it's a nice rule. 

I think the discussion around whether it belongs in core or as part of a plugin pack is connected to `no-unknown-animations` inclusion in https://github.com/stylelint/stylelint/issues/2079.

We might as well use this issue to discussion these two rules. (We agreed to group together rules with similar issues for easier discussion).

I think these two rules, arguably, fail to adhere to the following two [criteria](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#criteria-for-inclusion):

> Useful to the majority of users
> Has a clear and unambiguous finished state.

These rules are applicable to users who include their custom properties and at-keyframes definitions _within the same source that they are used_ i.e. those users who use methodologies similar to SuitCSS or CSS Modules. These rules aren't applicable if the custom properties and at-keyframes definitions are located _in sources separate from where they are used_, e.g. within an `animations.css` and `variables.css` files, (unless stylelint is used _after_ the sources have been processed and concatenated - an uncommon use-case?). 

I think the use of `unknown` creates an ambiguity (although, that's a bit tenuous as you could define `unknown` as "within the same source" in the docs) and I think the above outlines why these rules might not be useful to the majority of users.

As such, I think both these rules could be good candidates for great plugins. Ones that users of SuitCSS or CSS Modules-like methodologies can make use of.

As with each group of rules within https://github.com/stylelint/stylelint/issues/2079, I'm keen to hear counter arguments (or agreements). I think the exercise of discussing these fringe rules is worthwhile in itself as I think it'll help us to further refine our criteria moving into the new year.

**Comment by evilebottnawi** [comment]
@jeddy3 
> Useful to the majority of users

I think it would be useful for most users when the variables are supported by all popular browsers (http://caniuse.com/#feat=css-variables), even now it is not longer a small amount, also worth adding that this is part of the official `css`.

> sources separate from where they are used

I think the import directives analysis, not a difficult task, moreover, it can give a new breath to the new rules and improving linting (ast tree analysis is only part of lint).

> Has a clear and unambiguous finished state.

I do not agree, I would like to hear arguments :thinking:

**Comment by davidtheclark** [comment]
@evilebottnawi We already support rules related to custom properties, so I don't think browser support needs to affect the decision at this point. Also, I disagree that adding import directive analysis would be "not a difficult task". Sounds like it would add a lot of complexity. I'd be curious to see ideas about an implementation, though.

However, I don't think our scoping of known-unknown within a single stylesheet is grounds for removing the rules. These are simple rules, and as long as they are scoped to individual stylesheets I think they do have clear finished states. We had the original requests about them and we've had some follow-up on them, so it seems they are being used. I think we should keep them.

Thinking about the above makes me question the value of "useful to the majority of users" as a criterion. First, because we'll never have a sound way to judge this. Second, because plenty of rules might appeal to a significant group of users but not "the majority of users". I totally agree that we shouldn't be adding rules for individual's idiosyncratic whims, but "majority" isn't right. How about something more openly subjective and qualitative, like "Generally useful; not tied to idiosyncratic patterns"?

**Comment by jeddy3** [comment]
> Also, I disagree that adding import directive analysis would be "not a difficult task". Sounds like it would add a lot of complexity.

This was my feeling too.

> I'd be curious to see ideas about an implementation, though. 

Agreed. Happy to be proven wrong on this.

> However, I don't think our scoping of known-unknown within a single stylesheet is grounds for removing the rules. 

I see your point. Let's see if anymore of the team have opinions. Either way, clarify within the docs that "unknown" is scoped to the individual source would be useful.

>  First, because we'll never have a sound way to judge this.

That is very true.

> How about something more openly subjective and qualitative, like "Generally useful; not tied to idiosyncratic patterns"?

Sgtm.

> what do you think about the support new rules in core (custom-selectors, variables and other from new spec)

Yep, in-line with our design principles we should have plugins/rules for each specific construct. As each would be scoped to the source (or sources if import directive analysis is proven viable) I think they should have the following names:

- `no-unknown-custom-properties` ([ref](https://drafts.csswg.org/css-variables/#defining-variables))
- `no-unknown-custom-property-sets` ([ref](https://tabatkins.github.io/specs/css-apply-rule/#defining))
- `no-unknown-custom-selectors` ([ref](https://drafts.csswg.org/css-extensions/#custom-selectors))

**Comment by jeddy3** [comment]
As per:

> However, I don't think our scoping of known-unknown within a single stylesheet is grounds for removing the rules.

@dylang Would you like [to contribute](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md) this rule? I recommend using the [`no-unknown-animations`](https://github.com/stylelint/stylelint/tree/master/lib/rules/no-unknown-animations) code and docs as a blueprint.

To begin with, like `no-unknown-animations`, this rule will be scoped to each individual source. We can always investigate broadening this scope at a latter date.

Details:

- Name: `no-unknown-custom-properties`
- Description: "Disallow unknown custom properties"
- Extended description: "This rule considers custom properties defined within the same source to be known."
- Primary options: `true`
- Secondary options: none
- Message: `Unexpected unknown custom property`
- Section: "General / Sheet"

**Comment by hudochenkov** [comment]
Will it throw a warning?

```css
.parent {
	--color: red;
}
.child {
	color: var(--color);
}
```

It's a completely correct example for the following markup:

```html
<div class="parent">
	<div class="child">Text</div>
</div>
```

[Codepen](http://codepen.io/hudochenkov/pen/Rpowob?editors=1100).

I think this rule not that easy as `no-unknown-animations`, because `@keyframes` could be in the root node only. I think this rule could check the following cases only: custom properties declared in `:root` or in the same rule. We should _skip_ unknown custom properties if they were not found in `:root` or in the same node. Because the custom property could be declared in other rule. And this rescrictions deminish rule's value. It will never throw a warning ¯﻿\﻿_﻿(﻿ツ﻿)﻿_﻿/﻿¯

The following wouldn't considered warnings:

```css
.parent {
	--color: red;
}
.child {
	color: var(--color);
}
```

```css
.parent {
	--color: red;
}
.child {
	width: var(--width);
}
```

```css
:root {
	--color: red;
}
.parent {
	--width: 10px;
}
.child {
	width: var(--width);
}
```

```css
:root {
	--color: red;
}
.child {
	width: var(--width);
}
```

**Comment by davidtheclark** [comment]
I'm in agreement here with @hudochenkov ... sorry, I feel like I derailed the discussion above. This rule needs to be a plugin for now.

**Comment by jeddy3** [comment]
@hudochenkov Thanks for digging deeper into this. It's often the case that a rule that looks seemingly simple unravels when the details are thrashed out. 

Closing this issue as it's label as `type: plugin`. 

@dylang Feel free to [create a community plugin](https://stylelint.io/developer-guide/plugins/) for this feature. A plugin is the ideal place to experiment with ways to address the complexities that @hudochenkov outlined above.

**Comment by jonathantneal** [comment]
I’ve written a stylelint plugin to disallow usage of unknown custom properties.

https://github.com/csstools/stylelint-value-no-unknown-custom-properties

**Comment by jeddy3** [comment]
@jonathantneal Thanks! Looks great. I've created a PR, https://github.com/stylelint/stylelint/pull/3692, to add it to the list of plugins.

**Comment by silverwind** [comment]
I feel like a rule for undefined custom properties should be part of stylelint.

The main challenge I see is for stylelint to recognize all variable definitions from multiple files, but I guess with a new config system similar to ESLint flat, it should be a breeze to have rules that work on multiple files.

**Comment by jeddy3** [comment]
> I feel like a rule for undefined custom properties should be part of stylelint.

@silverwind Yeap, we agree. Ready to implement issue is https://github.com/stylelint/stylelint/issues/6361. I should've reopened this one, but I forgot we had it.

**Comment by silverwind** [comment]
Ah, yes that was the one. Google mislead me here 😉
